public/js/components/Sources.js

```diff
@@ -3,68 +3,261 @@
-          Isvg({ src: "images/globe.svg" }),
-          dom.span({ className: "label" }, domain),
-          dom.ul(null,
-            sourcesByDomain.get(domain).map(source => renderSource({
-              source, selectSource, selectedSource }))
-          )
-        );
-      })
+function createNode(name, path, contents) {
+  return [name, path, contents];
+}
+
+function createParentMap(tree) {
+  const map = new WeakMap();
+
+  function _traverse(subtree) {
+    if (nodeHasChildren(subtree)) {
+      for (let child of nodeContents(subtree)) {
+        map.set(child, subtree);
+        _traverse(child);
+      }
+    }
+  }
+
+  // Don't link each top-level path to the "root" node because the
+  // user never sees the root
+  nodeContents(tree).forEach(_traverse);
+  return map;
+}
+
+function getURL(source) {
+  try {
+    if (!source.get("url")) {
+      return null;
+    }
+
+    const url = new URL(source.get("url"));
+
+    // Filter out things like `javascript:<code>` URLs for now.
+    // Whitelist the protocols because there may be several strange
+    // ones.
+    if (url.protocol !== "http:" && url.protocol !== "https:") {
+      return null;
+    }
+
+    return url;
+  } catch (e) {
+    // If there is a parse error (which may happen with various
+    // internal script that don't have a correct URL), just ignore it.
+    return null;
+  }
+}
+
+function addToTree(tree, source) {
+  const url = getURL(source);
+  if (!url) {
+    return;
+  }
+
+  const parts = url.pathname.split("/").filter(p => p !== "");
+  const isDir = (parts.length === 0 ||
+                 parts[parts.length - 1].indexOf(".") === -1);
+  parts.unshift(url.host);
+
+  let path = "";
+  let subtree = tree;
+
+  for (let part of parts) {
+    const subpaths = nodeContents(subtree);
+    // We want to sort alphabetically, so find the index where we
+    // should insert this part.
+    let idx = subpaths.findIndex(subpath => {
+      return nodeName(subpath).localeCompare(part) >= 0;
+    });
+
+    // The node always acts like one with children, but the code below
+    // this loop will set the contents of the final node to the source
+    // object.
+    const pathItem = createNode(part, path + "/" + part, []);
+
+    if (idx >= 0 && nodeName(subpaths[idx]) === part) {
+      subtree = subpaths[idx];
+    } else {
+      // Add a new one
+      const where = idx === -1 ? subpaths.length : idx;
+      subpaths.splice(where, 0, pathItem);
+      subtree = subpaths[where];
+    }
+
+    // Keep track of the subpaths so we can tag each node with them.
+    path = path + "/" + part;
+  }
+
+  // Store the soure in the final created node.
+  if (isDir) {
+    setNodeContents(
+      subtree,
+      [createNode("(index)", source.get("url"), source)]
+    );
+  } else {
+    setNodeContents(subtree, source);
+  }
+}
+
+// This is inline because it's much faster. We need to revisit how we
+// load SVGs, at least for components that render them several times.
```

Comment:
I didn't do this prematurely; I was using inline-svg before but it clearly slowed down the tree because it does an AJAX request for every single SVG, even if it's the same one. The library does not support caching, so I implemented it myself but still hit another problem. I think when you expanded a node, the arrows would not be in the children for a split second and it would "flash" into appearance, bumping the text to the right.

Anyway, this was very straight-forward and immediately worked really well. We may want to look into a SVG->React conversion tool.